File: Java-Core/src/main/java/prompto/literal/Literal.java
package prompto.literal;

import prompto.error.PromptoError;
import prompto.expression.IExpression;
import prompto.runtime.Context;
import prompto.utils.CodeWriter;
import prompto.value.IValue;

public abstract class Literal<T extends IValue> implements IExpression {
	
	String text;
	T value;
	
	protected Literal(String text, T value) {
		this.text = text;
		this.value = value;
	}
	
	@Override
	public void toDialect(CodeWriter writer) {
		writer.append(text);
	}
	
	@Override
	public String toString() {
		return text;
	}
	
	public T getValue() {
		return value;
	}
	
	@Override
	public IValue interpret(Context context) throws PromptoError {
		return value;
	}
	
	@Override
	public int hashCode() {
		return value.hashCode();
	}
	
	@Override
	public boolean equals(Object obj) {
		if(obj instanceof Literal)
			return value.equals(((Literal<?>)obj).value);
		else
			return value.equals(obj);
	}
	
}

File: Java-CodeServer/src/main/resources/js/prompto/literal/Literal.js
function Literal(text, value) {
	this.text = text;
	this.value = value;
	return this;
}
	
Literal.prototype.toDialect = function(writer) {
    writer.append(this.text);
};

Literal.prototype.toString = function() {
	return this.text;
};

Literal.prototype.getValue = function() {
	return this.value;
};

Literal.prototype.interpret = function(context) {
	return this.value;
};

exports.Literal = Literal;
